iroh-rpc/src/client.rs

```diff
@@ -226,85 +214,118 @@ async fn dial(
         resource_id: String,
     }
 
     #[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
-    struct GetPayload {
+    struct GetResponse {
         response: String,
     }
 
-    struct Get;
-
-    // DIG! I am struggling
-    // I'm also not sure if I just don't know what's going on, or if this is a bad arrangement (or
-    // both)
-    // Trying to implement Factory for Get, implying that it will be the `Get` handler that is added
-    // to one of the test rpc clients. Namespace "a", method "get", to be added to the b_rpc_config
-    // using
-    // `b_rpc_config.with_namespace("a", |namespace| {
-    //     namespace.with_method("get", Get)
-    // })`
-    // Confused about some things that aren't compiling & also how we ensure that `handler::State<T>`
-    // is a String in this case, is there a way to connect the State concrete type with the
-    // implementation of Get?
-    // anyway, would love some imput!
     async fn get(
         state: handler::State<String>,
-        stream_id: Option<u64>,
+        _stream_id: Option<u64>,
         param: Vec<u8>,
     ) -> Result<Vec<u8>, RpcError> {
-        // not sure where to include that param needs to have DeserializeOwned trait. I keep
-        // googling in circles, which makes me think I'm coming from it at the wrong angle,
-        // would love some guidance.
-        let req: GetParams = deserialize_request(&param)?;
+        let _: GetRequest = deserialize_request(&param)?;
 
-        let bytes = serialize_response(GetPayload {
+        let bytes = serialize_response(GetResponse {
             response: state.clone(),
         })?;
         Ok(bytes)
     }
 
     #[tokio::test]
     async fn client_example() {
-        let state = handler::State::new("Wooo!".to_string());
-        let multiaddr: Multiaddr = "/memory/1234".parse().unwrap();
-        let peer_id: PeerId = "12D3KooWGQmdpzHXCqLno4mMxWXKNFQHASBeF99gTm2JR8Vu5Bdc"
-            .parse()
-            .unwrap();
-        let namespace = String::from("namespace");
-        let keypair = Keypair::generate_ed25519();
-
-        let mut client = Client::rpc_from_config(
+        let a_state = handler::State::new(());
+        let a_keypair = Keypair::generate_ed25519();
+        let a_peer_id = a_keypair.public().to_peer_id();
+        let a_addr: Multiaddr = "/memory/1234".parse().unwrap();
+
+        let b_state = handler::State::new("Wooo!".to_string());
+        let b_addr: Multiaddr = "/memory/4321".parse().unwrap();
+        let b_keypair = Keypair::generate_ed25519();
+        let b_peer_id = b_keypair.public().to_peer_id();
+
+        let (mut a_client, a_server) = rpc_from_config(
+            RpcConfig::new()
+                .with_swarm(swarm::new_mem_swarm(a_keypair))
+                .with_state(a_state)
+                // can add the addrs here, or after the client has already
+                // been constructed
+                .with_addr("b", b_addr.clone(), b_peer_id),
```

Comment:
a nice pattern in rust to solve this kind of thing is the builder pattern, that would look like this here

```rust
let (client, server) = RpcBuilder::new()
  .with_swarm(swarm::new_mem_swarm(a_keypair))
  .with_state(a_state)
  // can add the addrs here, or after the client has already
  // been constructed
  .with_addr("b", b_addr.clone(), b_peer_id)
  .build()
  .expect("failed to build rpc");
```